{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Deliverables:  40 total points\n",
    "\n",
    " - this notebook with inspection intro and results filled out for CoordinateMathSoln  4 points\n",
    " - seqCleaner  9 points\n",
    " - fastqParse  9 points\n",
    " - coordinateMathSoln 9 points\n",
    " - converter 9 points\n",
    " - Optional: Triad class rewrite (5 points extra credit)\n",
    " \n",
    " These must be submitted as the notebook with 4(or 5) completed cells, \n",
    "     \n",
    " Due: Monday April 17, 2023 11:55pm\n",
    "\n",
    "# Lab 2 – Manipulating Data Types (40 points)\n",
    "## Overview\n",
    "Last week we introduced the Python function input(), which is used to take in string data. This week we will use input() to take in string and numeric (float and integer) data. \n",
    "\n",
    "Remember that all data returned from input() is a string object, so this will mean that you need to convert any numeric data to their respective numeric objects. The exercises in this lab will give you practice manipulating various types of data that commonly arise in computational biology problems.\n",
    "\n",
    "We are again using jupyter for this assignment. Please save this notebook into your LAB02 folder. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence cleanup\n",
    "In this exercise, you will create a program to “clean up” a sequence of DNA by removing ambiguous bases (denoted by “N”) output from a sequencer. Your task is to create a Python program called seqCleaner that\n",
    " - asks for and collects a sequence of DNA using input()\n",
    " - removes the ambiguous parts of the sequence, outputs the “cleaned” sequence, replacing the ambiguous parts with a count in {}’s.\n",
    "For example, if I enter the sequence of DNA “AaNNNNNNGTC” (without quotes), the program will output:<br>\n",
    "AA{6}GTC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hints: \n",
    " - The input sequence is not guaranteed to be uppercase, but should be interpreted as though it is all uppercase.\n",
    " - Only the letters (A,C,G,T,N) will be included in the input.\n",
    " - Only the DNA characters (A,C,G,T) should remain after the cleanup. \n",
    " - The input may include, at most, one block of 'N' characters. \n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and specific line or block comments that explain the semantics of your implementation.\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs. This should be in the form of a  program level dosctring."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### seqCleaner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-12T23:59:39.453Z",
     "iopub.status.busy": "2022-01-12T23:59:39.449Z",
     "iopub.status.idle": "2022-01-12T23:48:05.846Z",
     "shell.execute_reply": "2022-01-12T23:48:05.857Z"
    }
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Akash Pandit (aspandit)\n",
    "# Group Members: Preeti Rajarathinam(prrajara), Delson Hays (dhhays), Natalie Cellucci(ncellucc), Riya Tiloda (rtiloda)\n",
    "\n",
    "'''\n",
    "Read a DNA string from user input and return a collapsed substring of embedded Ns to: {count}.\n",
    "\n",
    "Example: \n",
    "input: AaNNNNNNGTC\n",
    "output: AA{6}GTC\n",
    "\n",
    "Any lower case letters are converted to uppercase\n",
    "'''\n",
    "\n",
    "class DNAstring(str):  # gotta love inheritance where I can use string methods on this object B)\n",
    "    \"\"\"DNAstring class built off of the string class, DNAstring objects will have all the properties of strings + a length method which returns their length\n",
    "    and a purify method which uppercases a string and condenses any blocks of N\"\"\"\n",
    "    \n",
    "    def length(self):\n",
    "        '''return the length of the string'''\n",
    "        return len(self)\n",
    "    \n",
    "    def purify(self):\n",
    "        '''Return an upcased version of the string, collapsing a single run of Ns.\n",
    "        Assumptions:\n",
    "        - inputs consist solely of capital or lowercase A, G, C, T, or N\n",
    "        - only 1 \"block\" of Ns appear per string'''\n",
    "        self = self.upper()  # convert all string characters to uppercase \n",
    "        nCount = self.count('N')  # get how many Ns appear in the block of Ns, save that as nCount\n",
    "        nIndex = self.index('N')  # find the index of the first occurrance of N in the string, save that as nIndex\n",
    "\n",
    "        pureDNA = list(self.replace('N', ''))  # clear out the Ns and split up the string into a list (mutable), save that as pureDNA\n",
    "        pureDNA.insert(nIndex, f'{{{nCount}}}')  # insert '{n_count}' at the index where the Ns began in pureDNA (list)\n",
    "\n",
    "        return ''.join(pureDNA)  # return pureDNA joined together as a string\n",
    "\n",
    "\n",
    "\"\"\"Driver\"\"\"\n",
    "def main():\n",
    "    ''' Get user DNA data and clean it up.'''\n",
    "    data = input('DNA data?')\n",
    "    while (data):\n",
    "        thisDNA = DNAstring(data)  # get a raw string of dna\n",
    "        pureData = thisDNA.purify()  # run the purify method on that raw string, storing the purified string as pureData\n",
    "        print(pureData)\n",
    "        data = input('DNA data?')\n",
    "    \n",
    "main()  # run the driver"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence information parsing\n",
    "In this exercise, you will create a program to “parse” sequence name information from a single line of a FASTQ formatted file. Your task is to create a Python script called fastqParse that:\n",
    " - asks for and collects the seqname line of a FASTQ file using input()\n",
    " - parses out each field of the run information from the string and displays each of them on a new line\n",
    "For example, if I enter the FASTQ seqname line:<br> \n",
    "__*@EAS139:136:FC706VJ:2:2104:15343:197393*__\n",
    "then the program will output:<br>\n",
    "Instrument = EAS139<br>\n",
    "Run ID = 136<br>\n",
    "Flow Cell ID = FC706VJ<br>\n",
    "Flow Cell Lane = 2<br>\n",
    "Tile Number = 2104<br>\n",
    "X-coord = 15343<br>\n",
    "Y-coord = 197393\n",
    "### Hints: \n",
    " - The input string is guaranteed to have 7 fields. \n",
    " - The first character of the FASTQ seqname line is “@” and each field of the run information is separated by a colon“:”.  \n",
    " - A reasonable solution would be around 16 lines of code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain documentation/comments\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fastqParse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Instrument = EAS139\n",
      "Run ID = 136\n",
      "Flow Cell ID = FC706VJ\n",
      "Flow Cell Lane = 2\n",
      "Tile Lane = 2104\n",
      "X-coord = 15343\n",
      "Y-coord = 197393\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Akash Pandit (aspandit)\n",
    "# Group Members: Preeti Rajarathinam(prrajara), Delson Hays (dhhays), Natalie Cellucci(ncellucc), Riya Tiloda (rtiloda)\n",
    "\n",
    "'''\n",
    "Design a class FastqString with a method parse() which when given a seqname FASTQ file name to parse, will return information on the instrument used,\n",
    "the run ID, the flow cell ID, the flow cell lane, the tile lane, the X-coordinate, and the Y-coordinate formatted into print statements.\n",
    "\n",
    "Inputs: string formatted as \"@Instrument:Run ID:Flow Cell ID:Flow Cell Lane:Tile Lane:X-coord:Y-coord\"\n",
    "Outputs: a series of print messages outputted to console in the following format:\n",
    "\n",
    "Instrument = [Instrument]\n",
    "Run ID = [Run ID]\n",
    "Flow Cell ID = [Flow Cell ID]\n",
    "Flow Cell Lane = [Flow Cell Lane]\n",
    "Tile Lane = [Tile Lane]\n",
    "X-coord = [X-coord]\n",
    "Y-coord =[Y-cord]\n",
    "'''\n",
    "class FastqString(str):\n",
    "    '''FastqString: a class that inherits from the string class which describes a string object (a FASTQ file seqname)\n",
    "    that can be parsed into stdout with the additional method FastqString.parse()'''\n",
    "\n",
    "    def parse(self) -> None:\n",
    "        '''method parse(): parses its FastqString object, outputting the various descriptors of the string'''\n",
    "        seqname = self.split(':')  # splits the fastqstring into a list of its different parts along the colon symbols\n",
    "\n",
    "        # the following print statements take the strings in seqname and print them in their respective spots\n",
    "\n",
    "        print(\"Instrument =\", seqname[0][1:])  # prints the instrument name without the first char (@) to stdout\n",
    "        print(\"Run ID =\", seqname[1])  # prints the run id to stdout\n",
    "        print(\"Flow Cell ID =\", seqname[2])  # prints the flow cell id to stdout\n",
    "        print(\"Flow Cell Lane =\", seqname[3])  # prints the flow cell lane to stdout\n",
    "        print(\"Tile Lane =\", seqname[4])  # prints the tile lane to stdout\n",
    "        print(\"X-coord =\", seqname[5])  # prints the x coordinate to stdout\n",
    "        print(\"Y-coord =\", seqname[6])  # prints the y coordinate to stdout\n",
    "        pass\n",
    "    \n",
    "def main():\n",
    "    ''' Driver function, this contains code to  collect a seqname for a FASTQ file, turn it into a FastqString object,\n",
    "    and parse that object'''\n",
    "    stringToParse = FastqString(input(\"Please enter the seqname line for a FASTQ file: \"))  # prompts user for fastq file seqname\n",
    "    stringToParse.parse()  # parses the user's input and prints the results to stdout\n",
    "  \n",
    "\n",
    "main()  # executes main function"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein coordinates\n",
    "In this exercise, you will create a program that takes three sets of atomic coordinates, all provided on a **single line**. The program then calculates the bond lengths and angles. For this program, you can start with the Triad class (provided). Your task is to create a Python program called coordinateMathSoln that:\n",
    " - asks for and collects three sets of coordinates using input(), only use 1 line for this data !!\n",
    " - outputs the N-C and N-Ca bond lengths and the C-N-Ca bond angle with correct number of significant digits (see below)\n",
    " \n",
    "For example, if I enter the following coordinates (notice.. they are all on one line !!!) :<br>\n",
    "\n",
    "C = (39.447, 94.657, 11.824) N = (39.292, 95.716, 11.027) Ca = (39.462, 97.101, 11.465) <br>\n",
    "\n",
    "then the program will output the following three lines:<br>\n",
    "N-C bond length = 1.33<br>\n",
    "N-Ca bond length = 1.46<br>\n",
    "C-N-Ca bond angle = 124.0<br> \n",
    "\n",
    "(Note: make sure that the angle returned is in **degrees** !!)\n",
    "### Hints: \n",
    " - Each coordinate will contain only 3 numeric values. \n",
    " - Bond lengths are the distance between two points, which for points P and Q in 3-space,\n",
    "$(P_x , P_y , P_z )$ and $(Q_x , Q_y , Q_z )$ respectively, the distance between them is:\n",
    "\n",
    "\n",
    "$$\n",
    "\\begin {eqnarray}\n",
    "     \\|PQ\\| &=&\\sqrt{((P_x-Q_x)^2 + (P_y-Q_y)^2 + (P_z-Q_z)^2)} \\\\\n",
    "            &=&\\sqrt{\\sum^{i \\in x,y,z}{(P_i-Q_i)^2}} \\\\\n",
    "\\end{eqnarray}\n",
    "$$\n",
    "\n",
    " - Bond angles can be calculated from the dot product.\n",
    " \n",
    "Let’s say that we have three points in space labeled P, Q and R. We are interested in the angle at point Q that is made from line segments QP and QR.\n",
    "The dot product tells us (for standard vectors P and R) that:\n",
    "$$P \\cdot R=\\|P\\| \\|R\\| \\cos{\\theta}$$\n",
    "in this notation,$\\|P\\|$ refers to the length of vector P as a standard vector (assumed to begin at the origin (0,0,0) ).\n",
    "We can then see that the angle between vectors P and R can be found by:\n",
    "$$\\cos{\\theta}=\\frac{P \\cdot R} {\\|P\\| \\|R\\|}$$\n",
    "\n",
    "We can calculate the dot product using the sum of products of the vector components:\n",
    "$$P \\cdot R = \\sum^{i \\in x,y,z}{P_i R_i} $$\n",
    "\n",
    "Now, to find vector P in standard form, we need to remember that QP starts at Q, so we need to place the origin at Q and find out where P is in that new space. We do that by subtracting the components of Q from P.\n",
    "Putting all of this together, we get:\n",
    "$$\\theta=\\cos^{-1}{\\frac{\\sum^{i \\in x,y,z}{(P_i - Q_i) (R_i - Q_i) }} {\\|QP\\| \\|QR\\|}}$$\n",
    "Remember, $\\theta$ is in radians.\n",
    "\n",
    "Below I have given you a class (Triad) with methods to calculate dot products (dot), dot products of translated vectors (ndot), distances (dPQ, dPR, dPR) and angles in radians (angleP, angleQ and angleR) for each of the three points in a Triad object.\n",
    "A reasonable solution for this exercise involves around 12 lines of additional code excluding comments.\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Introduction\n",
    "\n",
    "For this weeks inspection please focus on coordinateMathSoln. Consider the design choices that you made, how you are dealing with the single line of input data, how you are doing the data conversions, and how you convert to degrees.\n",
    "\n",
    "For this program, my main work was in designing the main function. \n",
    "\n",
    "Above all, I want my code to be concise and readable. I started this by separating my code into blocks that made sense, with each block being\n",
    "explained by its own single line comment(s).\n",
    "\n",
    "First, I just gathered user input and stored it in a variable. There's a problem here, this data is messy and needs to be parsed into the set of 3 tuples it should be in.\n",
    "\n",
    "Second, I parse the data. By comparing every character in the input string to a string of whitelisted characters, I recreate a list of the characters that pass the whitelist and are in both strings through a generator object with its iterables joined together into a string.\n",
    "This string is then parsed through list comprehension, where the string is first stripped of excess whitespace then split into a list of its numbers, which then the numbers are converted into tuples from strings through that list comprehension.\n",
    "\n",
    "Third, the point variables are assigned based on their respective thirds of the list of coordinates formed in part 2. These point tuples are then passed as arguments to initialize a Triad object. \n",
    "\n",
    "Fourth, a radian to degree conversion multiplier is defined for use later.\n",
    "\n",
    "Fifth and lastly, the distance and angle methods provided by the Triad class are implemented and their outputs formatted into strings. as Nitrogen was the central atom and occupied the q parameter slot, the .angleQ() method was chosen to represent the bond angle."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### coordinateMathSoln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N-C bond length = 33.67\n",
      "N-Ca bond length = 336.75\n",
      "C-N-Ca bond angle = 180.0\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Akash Pandit (aspandit)\n",
    "# Group Members: Preeti Rajarathinam(prrajara), Delson Hays (dhhays), Natalie Cellucci(ncellucc), Riya Tiloda (rtiloda)\n",
    "\n",
    "'''\n",
    "This program defines a class Triad with methods to get distance between 2 3-D points and the angle between 3 3-D points.\n",
    "A driver function main is also written to parse user input to be used as parameters in a Triad object for manipulation by\n",
    "its classes. \n",
    "\n",
    "A proper input would look like: 'C = (1x, 1y, 1z) N = (2x, 2y, 2z) Ca = (3x, 3y, 3z)'\n",
    "with the Nxyz parameters representing coordinates for each atom C, N, or Ca.\n",
    "\n",
    "A proper output, assuming N as the angle point, would display:\n",
    "'N-C bond length = {float}'\n",
    "'N-Ca bond length = {float}'\n",
    "'C-N-Ca bond angle = {float}' \n",
    "'''\n",
    "\n",
    "import math\n",
    "class Triad :\n",
    "    \"\"\"\n",
    "    Calculate angles and distances among a triad of points.\n",
    " \n",
    "    Author: David Bernick\n",
    "    Date: March 21, 2013\n",
    "    Points can be supplied in any dimensional space as long as they are consistent.\n",
    "    Points are supplied as tupels in n-dimensions, and there should be three\n",
    "    of those to make the triad. Each point is positionally named as p,q,r\n",
    "    and the corresponding angles are then angleP, angleQ and angleR.\n",
    "    Distances are given by dPQ(), dPR() and dQR()\n",
    " \n",
    "    Required Modules: math\n",
    "    initialized: 3 positional tuples representing Points in n-space\n",
    "             p1 = Triad( p=(1,0,0), q=(0,0,0), r=(0,1,0) )\n",
    "    attributes: p,q,r the 3 tuples representing points in N-space\n",
    "    methods:  angleP(), angleR(), angleQ() angles measured in radians\n",
    "          dPQ(), dPR(), dQR() distances in the same units of p,q,r\n",
    " \n",
    "    \"\"\"\n",
    " \n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad. \n",
    "        \n",
    "        Example construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ). \n",
    "        \"\"\"\n",
    "        self.p = p\n",
    "        self.q = q\n",
    "        self.r = r\n",
    "        \n",
    "# private helper methods\n",
    "    def d2 (self,a,b) : # calculate squared distance of point a to b\n",
    "        return float(sum((ia-ib)*(ia-ib)  for  ia,ib in zip (a,b)))\n",
    "    \n",
    "    def dot (self,a,b) : # dotProd of standard vectors a,b\n",
    "        return float(sum(ia*ib for ia,ib in zip(a,b)))\n",
    "    \n",
    "    def ndot (self,a,b,c) : # dotProd of vec. a,c standardized to b\n",
    "        return float(sum((ia-ib)*(ic-ib) for ia,ib,ic in zip (a,b,c)))\n",
    "    \n",
    "# calculate lengths(distances) of segments PQ, PR and QR\n",
    "    def dPQ (self):\n",
    "        \"\"\" Provides the distance between point p and point q \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.q))\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Provides the distance between point p and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.p,self.r))\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Provides the distance between point q and point r \"\"\"\n",
    "        return math.sqrt(self.d2(self.q,self.r))\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Provides the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.q,self.p,self.r) / \n",
    "                         math.sqrt(self.d2(self.q,self.p) * self.d2(self.r,self.p)))\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Provides the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.q,self.r) /\n",
    "                         math.sqrt(self.d2(self.p,self.q) * self.d2(self.r,self.q)))\n",
    " \n",
    "    def angleR (self) :\n",
    "        \"\"\" Provides the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        return math.acos(self.ndot(self.p,self.r,self.q) /\n",
    "                         math.sqrt(self.d2(self.p,self.r) * self.d2(self.q,self.r)))\n",
    "\n",
    "def main():\n",
    "    '''Driver function- gather user input for coordinates, clean the data, then parse the distances and angles and output to stdout'''\n",
    "    # get user input\n",
    "    coords = input(\"Please input your coordinates!\")\n",
    "\n",
    "    # clean user input and turn it into a list of the floats\n",
    "    charsToInclude = '0123456789. '\n",
    "    coords = ''.join(char for char in coords if char in charsToInclude)  # rewrite string with only chars in charsToInclude\n",
    "    coords = [float(num) for num in coords.strip().split()]  # split into list of the nums and convert them to floats\n",
    "\n",
    "    # assign points to input to point vars & initialize Triad object with them\n",
    "    point1, point2, point3 = tuple(coords[:3]), tuple(coords[3:6]), tuple(coords[6:9])\n",
    "    triad = Triad(point1, point2, point3)\n",
    "\n",
    "    radToDeg = 180 / math.pi  # conversion multiplier from radians to degrees\n",
    "\n",
    "    print('N-C bond length =', f'{triad.dPQ():.2f}')  # prints bond length between N and C to 3 sig decimals\n",
    "    print('N-Ca bond length =', f'{triad.dQR():.2f}')  # prints bond length between N and Ca to 3 sig decimals\n",
    "    print('C-N-Ca bond angle =', round(triad.angleQ() * radToDeg, 0))  # prints the bond angle at Nitrogen\n",
    "\n",
    "\n",
    "main()  # execute driver"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "## Inspection Results \n",
    "(Group, Findings and your Responses to those findings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra credit (5 points): Rewrite the Triad class.  \n",
    "For extra-credit, provide a direct replacement for the Triad class. The external methods that calculate angles, distances, and points (tuples) p,q and r must be maintained such that either version of the Triad class can be used.\n",
    "\n",
    "You could use the cosine law to calculate angles instead of the dot product.  You might make use of the numpy module. You might recode each of the methods to avoid using zip. You might consider using list iterations.\n",
    "Your Triad replacement must reimplement all of Triad public function, without using zip and without being a trivial rewrite. Your implementation need not be as compact as the current implementation, and it needs to be correct and fully documented to receive full credit. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Triad rewrite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22680.0 -22680.0\n",
      "-24948.0 -24948.0\n",
      "-249480.0 -249480.0\n",
      "N-C bond length = 33.67\n",
      "N-Ca bond length = 336.75\n",
      "C-N-Ca bond angle = 180.0\n",
      "\n",
      "Hypothetical N-C-Ca bond angle = 180.0\n",
      "Hypothetical N-Ca-C bond angle = 180.0\n",
      "Shows sum of angles = 180deg: 540.0\n",
      "(above 2 print statements are just proof of concept of triad.lawOfCosForAngle())\n"
     ]
    }
   ],
   "source": [
    "from math import acos, dist, pi  # why reinvent the wheel skull emoji\n",
    "\n",
    "class Triad:\n",
    "    def __init__(self,p,q,r) :\n",
    "        \"\"\" Construct a Triad.  \n",
    "        \n",
    "        Example object construction:\n",
    "            p1 = Triad( p=(1.,0.,0.), q=(0.,0.,0.), r=(0.,0.,0.) ).\n",
    "        \"\"\"\n",
    "        self.p = p  # first point\n",
    "        self.q = q  # second point\n",
    "        self.r = r  # third point\n",
    "    def dPQ (self):\n",
    "        \"\"\" Return the distance between point p and point q \"\"\"\n",
    "        return dist(self.p, self.q)  # calculates distance with math.dist (distance formula func)\n",
    "    \n",
    "    def dPR (self):\n",
    "        \"\"\" Return the distance between point p and point r \"\"\"\n",
    "        return dist(self.p, self.r)  # calculates distance with math.dist (distance formula func)\n",
    "    \n",
    "    def dQR (self):\n",
    "        \"\"\" Return the distance between point q and point r \"\"\"\n",
    "        return dist(self.q, self.r)  # calculates distance with math.dist (distance formula func)\n",
    "\n",
    "    def lawOfCosForAngle(self, a, b, c):\n",
    "        \"\"\"executes the law of cosines solving for the angle (designated A)\n",
    "\n",
    "        law of cosines: c = sqrt(a**2 + b**2 - 2ab * cos(A))\n",
    "        c**2 = a**2 + b**2 - 2ab * cos(A) \n",
    "        c**2 - a**2 - b**2 = -2ab * cos(A) \n",
    "        cos(A) = (c**2 - a**2 - b**2) / (-2ab) \n",
    "        A = arccos((c**2 - a**2 - b**2) / (-2ab))\"\"\"\n",
    "\n",
    "        # round functions are implemented with large significant figure params to account for straight line molecular geometries like in CO2,\n",
    "        # the 180 degree bond angle in a perfect world would lead to either acos(1) or acos(-1), right at the edge of its domain, but due to\n",
    "        # floating point imprecision the value can sometimes be very slightly outside the domain of acos(), resulting in an error. The significant\n",
    "        # figure parameter in the round functions provides healthy precision without being so precise as to allow for floating point error and\n",
    "        # jump outside of the domain of acos()\n",
    "        return acos(round((c**2 - a**2 - b**2), 10) / round(-2*a*b, 10)) \n",
    "        # returns angle in radians solving for angle with law of cosines (work shown in docstring)\n",
    "    \n",
    "    def angleP (self) :\n",
    "        \"\"\" Return the angle made at point p by segments pq and pr (radians). \"\"\"\n",
    "        return self.lawOfCosForAngle(self.dPQ(), self.dPR(), self.dQR())\n",
    "        # uses my method lawOfCosForAngle to calculate the angle\n",
    "    \n",
    "    def angleQ (self) :\n",
    "        \"\"\" Return the angle made at point q by segments qp and qr (radians). \"\"\"\n",
    "        return self.lawOfCosForAngle(self.dPQ(), self.dQR(), self.dPR())\n",
    "        # uses my method lawOfCosForAngle to calculate the angle\n",
    "        \n",
    "    def angleR (self) :\n",
    "        \"\"\" Return the angle made at point r by segments rp and rq (radians). \"\"\"\n",
    "        return self.lawOfCosForAngle(self.dPR(), self.dQR(), self.dPQ())\n",
    "        # uses my method lawOfCosForAngle to calculate the angle\n",
    "        \n",
    "        \n",
    "def main() -> None:\n",
    "    '''Driver function- gather user input for coordinates, clean the data, then parse the distances and angles and output to stdout\n",
    "    .angleP() and .angleR() are also implemented in this example to show proof of concept.'''\n",
    "    coords = input(\"Please input your coordinates!\")\n",
    "    charsToInclude = '0123456789. '\n",
    "    coords = ''.join(char for char in coords if char in charsToInclude)  # rewrite string with only chars in charsToInclude\n",
    "    coords = [float(num) for num in coords.strip().split()]  # split into list of the nums and convert them to floats\n",
    "\n",
    "    # assign points to input to point vars & \n",
    "    point1, point2, point3 = tuple(coords[:3]), tuple(coords[3:6]), tuple(coords[6:9])\n",
    "    triad = Triad(point1, point2, point3)\n",
    "\n",
    "    # creates triad object with Carbon, Nitrogen, and Calcium coordinates respectively\n",
    "    radToDeg = 180 / pi  # defines conversion factor from radians to degrees\n",
    "    angleQ = round(triad.angleQ() * radToDeg, 0)  # converts .angleQ() to degrees then rounds it to 1 significant decimal\n",
    "    angleP = round(triad.angleP() * radToDeg, 0)  # converts .angleP() to degrees then rounds it to 1 significant decimal\n",
    "    angleR = round(triad.angleR() * radToDeg, 0)  # converts .angleR() to degrees then rounds it to 1 significant decimal\n",
    "\n",
    "    print('N-C bond length =', f'{triad.dPQ():.2f}')  # prints N-C bond length  to 2 sig decimals\n",
    "    print('N-Ca bond length =', f'{triad.dQR():.2f}')  # prints N-Ca bond length to 2 sig decimals\n",
    "    print('C-N-Ca bond angle =', angleQ)  # prints the bond angle at Nitrogen\n",
    "    \n",
    "    print('\\nHypothetical N-C-Ca bond angle =', angleP)  # prints bond angle at Carbon\n",
    "    print('Hypothetical N-Ca-C bond angle =', angleR)  # prints bond angle at Calcium\n",
    "    print('Shows sum of angles = 180deg:', angleP + angleQ + angleR)  \n",
    "    # shows how this matches that sum(Triangle Angles) = 180deg\n",
    "    print('(above 2 print statements are just proof of concept of triad.lawOfCosForAngle())')  # explains the above\n",
    "\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Codon tables and amino acid letter converters\n",
    "In this exercise, you will create a Python solution that uses mappings to convert sequence information between different amino acid representations.  This includes the 3-letter codon code (RNA and DNA), the one letter amino acid code and the 3-letter amino acid code.\n",
    "The program will use different dictionaries that represent :\n",
    "codon tables – one for DNA and the other for RNA, and\n",
    "amino acid letter representation converters. \n",
    "\n",
    "Your task is to create a solution that\n",
    "asks for and collects a single input string using input()\n",
    "parses the string, looks up the information in the appropriate dictionary, and outputs the correct conversion\n",
    "For example:<br>\n",
    "\n",
    "if I enter “ATG” (without quotes), then the program will output:<br>\n",
    "ATG = MET<br>\n",
    " \n",
    "if I enter “UAG” (without quotes), then the program will output:<br>\n",
    "UAG = ---<br>\n",
    "if I enter “E” (without quotes), then the program will output:<br>\n",
    "E = GLU<br>\n",
    "if I enter “Asp” (without quotes), then the program will output:<br>\n",
    "ASP = D<br>\n",
    "### Hints: \n",
    " - The program might not get a valid codon.  In that case, it should output 'unknown'.  You can use the dictionary method 'get' and include a default_value to handle the 'unknown' case. Remeeber our discussion of how  the _or_ function is a short-circuit operator.\n",
    " \n",
    "To get full credit on this assignment, your program needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain docstrings and line comments (using #)\n",
    " - Include any assumptions or design decisions you made in writing your code\n",
    " - Include an overview describing what your program does with expected inputs and outputs as a program level docstring\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### converter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unknown\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Akash Pandit\n",
    "# Group Members: Preeti Rajarathinam(prrajara), Delson Hays (dhhays), Natalie Cellucci(ncellucc), Riya Tiloda (rtiloda)\n",
    "\n",
    "'''\n",
    "This program contains various dictionaries displaying the key:value relationships between amino acid representations\n",
    "and codons, both DNA and RNA. A driver function main takes a user input and tests whether it is a key in any of the\n",
    "given dictionaries and returns its appropriate value, or 'unknown' if nothing matches.\n",
    "\n",
    "This is done through a series of conditional statements checking if an input is in a dictionary's keys, and returning\n",
    "the appropriate value if so.\n",
    "\n",
    "Some example inputs and their outputs in format:\n",
    "input = output\n",
    "\n",
    "C = CYS\n",
    "GLY = G\n",
    "UUU = PHE\n",
    "ATC = Ile\n",
    "0 = unknown\n",
    "baller = unknown\n",
    "\n",
    "The names of short_AA, long_AA, and RNA_codon_table changed to shortAA, longAA, and rnaCodonTable respectively\n",
    "to match this class' request to use camelCase naming conventions.\n",
    "'''\n",
    "shortAA = {\n",
    "            'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "            'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "            'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "            'ALA': 'A', 'VAL': 'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'\n",
    "            }  # maps 3 char amino acid acronyms to 1 char acronyms\n",
    "\n",
    "longAA = {value:key for key,value in shortAA.items()}  # maps 1 char amino acid acronyms to 3 char acronyms\n",
    "\n",
    "rnaCodonTable = {  # maps RNA codons to their respective amino acids, with stop codons designated as ---\n",
    "# Second Base\n",
    "# U             C             A             G\n",
    "#U\n",
    "'UUU': 'Phe', 'UCU': 'Ser', 'UAU': 'Tyr', 'UGU': 'Cys',\n",
    "'UUC': 'Phe', 'UCC': 'Ser', 'UAC': 'Tyr', 'UGC': 'Cys',\n",
    "'UUA': 'Leu', 'UCA': 'Ser', 'UAA': '---', 'UGA': '---',\n",
    "'UUG': 'Leu', 'UCG': 'Ser', 'UAG': '---', 'UGG': 'Trp',\n",
    "#C \n",
    "'CUU': 'Leu', 'CCU': 'Pro', 'CAU': 'His', 'CGU': 'Arg',\n",
    "'CUC': 'Leu', 'CCC': 'Pro', 'CAC': 'His', 'CGC': 'Arg',\n",
    "'CUA': 'Leu', 'CCA': 'Pro', 'CAA': 'Gln', 'CGA': 'Arg',\n",
    "'CUG': 'Leu', 'CCG': 'Pro', 'CAG': 'Gln', 'CGG': 'Arg',\n",
    "#A\n",
    "'AUU': 'Ile', 'ACU': 'Thr', 'AAU': 'Asn', 'AGU': 'Ser',\n",
    "'AUC': 'Ile', 'ACC': 'Thr', 'AAC': 'Asn', 'AGC': 'Ser',\n",
    "'AUA': 'Ile', 'ACA': 'Thr', 'AAA': 'Lys', 'AGA': 'Arg',\n",
    "'AUG': 'Met', 'ACG': 'Thr', 'AAG': 'Lys', 'AGG': 'Arg',\n",
    "#G\n",
    "'GUU': 'Val', 'GCU': 'Ala', 'GAU': 'Asp', 'GGU': 'Gly',\n",
    "'GUC': 'Val', 'GCC': 'Ala', 'GAC': 'Asp', 'GGC': 'Gly',\n",
    "'GUA': 'Val', 'GCA': 'Ala', 'GAA': 'Glu', 'GGA': 'Gly',\n",
    "'GUG': 'Val', 'GCG': 'Ala', 'GAG': 'Glu', 'GGG': 'Gly'\n",
    "}\n",
    "dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}  \n",
    "# same as rnaCodonTable but with T instead of U\n",
    "\n",
    "def main() -> None:\n",
    "    '''Driver function, prompts user for input then searches for that input in various dictionaries to output\n",
    "    the input's codon or amino acid equivalent, printing unknown if nothing is found.'''\n",
    "    inp = input('Search for your amino acid or codon here! ')  # prompt user to search for codon/AA\n",
    "\n",
    "    if inp.upper in longAA.keys():\n",
    "        print(inp.upper(), '=', longAA[inp.upper()].upper())  # the input was found in longAA\n",
    "\n",
    "    elif inp.upper() in shortAA.keys():\n",
    "        print(inp.upper(), '=', shortAA[inp.upper()].upper())  # the input was found in shortAA\n",
    "\n",
    "    elif inp.upper() in rnaCodonTable.keys():\n",
    "        print(inp.upper(), '=', rnaCodonTable[inp.upper()].upper())  # the input was found in rnaCodonTable\n",
    "\n",
    "    elif inp.upper() in dnaCodonTable.keys():\n",
    "        # for DNA inputs without T, they would be pulled from the RNA Codon Table as they would be identical to the DNA codon\n",
    "        print(inp.upper(), '=', dnaCodonTable[inp.upper()].upper())  # the input has a T and was found in dnaCodonTable\n",
    "\n",
    "    else:\n",
    "        print('unknown')  # the input was not found in any of the dictionaries\n",
    "    \n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit your code and answers\n",
    " \n",
    "Important: please save your work before logging out of the computer. This will ensure that you have a copy of your work and you will avoid having to redo everything in the  event that something happens with the lab computers. The three recommended solutions (pick one) at this point are to:\n",
    " - Email your code file to yourself\n",
    " - Copy your code file to a pen drive\n",
    " - Save your work on your own computer if you are using Anaconda ( or equiv)\n",
    "\n",
    "Submit the notebook created as part of this lab assignment, with cells completed:\n",
    " \n",
    " - this notebook file, complete with inspection introduction and inspection for coordMathSoln (4 points)\n",
    " - seqCleaner 9 points\n",
    " - fastqParse 9 points\n",
    " - coordinateMathSoln 9 points\n",
    " - converter 9 points\n",
    " - triad Triad rewrite (optional) 5 points\n",
    " \n",
    "Important: to get full credit on this lab assignment, each of the code cells you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Contain proper docstrings and appropriate line style comments (#)\n",
    " \n",
    "Congratulations, you finished your second lab assignment!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
