{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Flow Control\n",
    "\n",
    "\n",
    " - Due: April 24, 2023 at 11:55 pm, submit in Canvas\n",
    " - program filename: proteinParams.py \n",
    " - total points possible: 44\n",
    " - extra-credit possible: 6\n",
    " - submit two files:\n",
    "     - Lab03.ipynb (the notebook), and\n",
    "     - proteinParams.py (a python file that contains your code)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building and Testing\n",
    "For this lab, you will submit two files: the notebook with your inspection info and results, and file named proteinParams.py. You can do all of your work from within jupyter, then copy/paste your program to a new text file using your favorite text editor or python IDE. \n",
    "Test that program file !! On Mac systems you will use terminal, and on windows systems you will likely use cmd. You will then navigate to the directory where your proteinParam.py lives, then execute: <br>\n",
    "python3 proteinParam.py <br>\n",
    "or <br>\n",
    "python proteinParam.py <br>\n",
    "## Protein parameters\n",
    "In this exercise, youâ€™ll create a Python program to calculate the physical-chemical properties of a protein sequence similar to what ProtParam outputs. Your task is to develop the ProteinParam class included below in template form. The program includes the methods that you will need, along with the completed \"main\" that does all of the input and output.\n",
    "\n",
    "When testing, you will type the protein sequence, hit return, the program will respond with the required output.  You can then enter a new protein string and a new analysis will be presented.  When you are finished, type ctrl-D to signal the end (this is done by holding the control key down as you type the letter D).\n",
    "\n",
    "Your program will read in the protein sequence and print out the:\n",
    "\n",
    " - number of amino acids and total molecular weight,\n",
    " - Molar extinction coefficient and Mass extinction coefficient,\n",
    " - theoretical isoelectric point (pI), and\n",
    " - amino acid composition\n",
    " \n",
    "For example, if I enter the protein sequence:\n",
    "VLSPADKTNVKAAW \n",
    "\n",
    "then the program should output:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "Number of Amino Acids: 14\n",
    "Molecular Weight: 1499.7\n",
    "molar Extinction coefficient: 5500.00\n",
    "mass Extinction coefficient: 3.67\n",
    "Theoretical pI: 9.88\n",
    "Amino acid composition:\n",
    "A = 21.43%\n",
    "C = 0.00%\n",
    "D = 7.14%\n",
    "E = 0.00%\n",
    "F = 0.00%\n",
    "G = 0.00%\n",
    "H = 0.00%\n",
    "I = 0.00%\n",
    "K = 14.29%\n",
    "L = 7.14%\n",
    "M = 0.00%\n",
    "N = 7.14%\n",
    "P = 7.14%\n",
    "Q = 0.00%\n",
    "R = 0.00%\n",
    "S = 7.14%\n",
    "T = 7.14%\n",
    "V = 14.29%\n",
    "W = 7.14%\n",
    "Y = 0.00%\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hints:\n",
    "\n",
    "The input sequence is not guaranteed to be uppercase and might contain unexpected characters.  \n",
    "Only count the following (A, C, D, E, F, G, H, I, L, K, M, N, P, Q, R, S, T, V, Y, W) or the lower-case equivalents, and ignore anything else. Math details are included below.\n",
    "\n",
    "\n",
    "To get full credit on this assignment, your code needs to:\n",
    "\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Include a docstring overview about what your program is designed to do with expected inputs and outputs\n",
    " - include docstrings for every class, method within that class\n",
    " - Include any assumptions or design decisions you made in writing your code as # comments\n",
    " - Contain in-line comments using #-style where appropriate. Make sure to fix the template (below) to conform.\n",
    " \n",
    " - Submit your proteinParams.py file and the notebook using Canvas.\n",
    "\n",
    "Congratulations, you finished your third lab assignment!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design specification \n",
    "\n",
    "### \\_\\_init\\_\\_\n",
    "\n",
    "There are a number of ways to design this.  Your \\_\\_init\\_\\_ method could save an attribute which is just the input string. A more effective solution would compute and save the aaComposition dictionary. All of the protein parameter methods can operate very efficiently using a dictionary. Use of an aaComposition here will save you quite a bit of work.\n",
    "\n",
    "### aaCount()\n",
    "\n",
    "This method will return a single integer count of valid amino acid characters found. Do not assume that this is the length of the input string, since you might have spaces or invalid characters that are required to be ignored.\n",
    "\n",
    "### aaComposition() - 4 points\n",
    "\n",
    "This method is to return a dictionary keyed by single letter Amino acid code, and having associated values that are the counts of those amino acids in the sequence. Make sure to include all 20 amino acids.  Proper amino acids that are not represented in the sequence should have a value of zero. Note: if you have already calculated a composition dictionary in __init__, then just return that dictionary here.\n",
    "\n",
    "### molecularWeight() - 8 points\n",
    "\n",
    "This method calculates the molecular weight (MW) of the protein sequence. If we have the composition of our protein, this is done by summing the weights of the individual Amino acids and excluding the waters that are released with peptide bond formation. \n",
    "$$MW_{H_{2}O}+\\sum^{aa}N_{aa}(MW_{aa}-MW_{H_{2}O})$$\n",
    "\n",
    "### \\_charge\\_(pH) -- 10 points\n",
    "\n",
    "This method calculates the net charge on the protein at a specific pH (specified as a parameter of this method). The method is used by the pI method. I have marked it with the single _ notation to advise future users of the class that this is not part of the defined interface and it just might change.\n",
    "\n",
    "If we have the composition of our protein, we can then calculate the net charge at a particular pH, using the pKa of each charged Amino acid and the Nterminus and Cterminus.\n",
    "\n",
    "$$ netCharge = \\left[\\sum^{aa=(Arg,Lys,His,Nterm}N_{aa}\\frac{10^{pKa(aa)}}{10^{pKa(aa)}+10^{pH}}\\right] - \n",
    "               \\left[\\sum^{aa=(Asp,Glu,Cys,Tyr,Cterm}N_{aa}\\frac{10^{pH}}{10^{pKa(aa)}+10^{pH}}\\right]\n",
    "$$\n",
    "\n",
    "I have provided pKa tables for each AA, and the pKa for the N-terminus and C-terminus.\n",
    "\n",
    "### pI() - 10 points\n",
    "\n",
    "The theoretical isolelectric point can be estimated by finding the  particular pH that yields a neutral net Charge (close to 0). There are a few ways of doing this, but the simplest might be to iterate over all pH values to find the one that is closest to 0. Doing this by hand is painful, but its not that bad to do computationally.  Remember that we want to find the  best pH, accurate to 2 decimal places.\n",
    "\n",
    "#### extra credit (3 points) for pI() method\n",
    "\n",
    "Another way of doing the pI calculation would use a binary search over the pH range. This works because we expect a single zero crossing to exist in the range, and the function will be well behaved across the range of charge() as a function of pH (0-14 range). You then make the algorithm operate to any specified precision using an optional parameter (set the default parameter: precision = 2).\n",
    "\n",
    "### molarExtinction() - 8 points\n",
    "\n",
    "The extinction coefficient indicates how much light a protein absorbs at a certain wavelength. It is useful to have an estimation of this coefficient for measuring a protein with a spectrophotometer at a wavelength of 280nm. It has been shown by Gill and von Hippel that it is possible to estimate the molar extinction coefficient of a protein from knowledge of its amino acid composition alone. From the molar extinction coefficient of tyrosine, tryptophan and cystine at a given wavelength, the extinction coefficient of the native protein in water can be computed using the following equation.\n",
    "\n",
    "$$E = N_Y E_Y + N_W E_W + N_C E_C$$\n",
    "where:\n",
    "\n",
    "- $N_Y$ is the number of tyrosines, $N_W$ is the number of tryptophans, $N_C$ is the number of cysteines,\n",
    "- $E_Y$ , $E_W$, $E_C$ are the extinction coefficients for tyrosine, tryptophan, and cysteine respectively.\n",
    "\n",
    "I have supplied the molar extinction coefficients at 280nm for each of these residues in a dictionary(aa2abs280) in the program template.\n",
    "\n",
    "Note that we will assume for this exercise that all Cysteine residues are represented as Cystine. Under reducing conditions, Cystine does not form however and Cysteine residues do not contribute to absorbance at 280nm.\n",
    "\n",
    "### massExtinction()\n",
    "\n",
    "We can calculate the Mass extinction coefficient from the Molar Extinction coefficient by dividing by the molecularWeight of the corresponding protein.\n",
    "\n",
    "#### extra credit for molarExtinction() and massExtinction() - 3 points\n",
    "\n",
    "As mentioned above, we are assuming that all Cysteine residues are present as Cystine.  Provide an optional parameter to both molarExtinction() and massExtinction() to calculate these values assuming reducing conditions. Use an optional parameter with a default of True (Cystine=True) to allow evaluation of both molar and mass extinction under both oxidizing (default) and reducing conditions."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "# Inspection info\n",
    "\n",
    "Describe all of the information that your inspection team needs to know to understand your design and implementation. Examples:\n",
    "- How did you save the essential data attribute in objects of ProteinParam ?\n",
    "\n",
    "my constructor initialized 2 instance variables: self.protein and self.aaComposition_. self.protein uses a list comprehension to build a list of all the characters in the protein input parameter but only including them if they are in the list of keys for the dictionary mapping amino acids to their molecular weights, aa2mw. The amino acid list is then joined together into a string and capitalized for future convenience. self.aaComposition_ uses dictionary comprehension to loop through every character in self.protein and set the amino acid as the key and its count in self.protein as the value. No need to handle duplicate values, as they are just rewritten by the same identical value.\n",
    "\n",
    "- How did you make use of that save attribute for each of your methods ?\n",
    "\n",
    "in self.aaCount(), since self.protein was already cleaned in the constructor, I only had to return its length.\n",
    "in self._charge_(), I iterate over every value in self.protein to check for charged amino acids. This information is later used in self.pI().\n",
    "in self.aaComposition(), since self.aaComposition_ was already defined in the constructor, I just return that value.\n",
    "in self.molarExtinction(), when implementing the molar extinction equation I can just run the str.count() method on self.protein to get the number of each relevant amino acid, data that is also used in self.massExtinction().\n",
    "self.molWeight(), like self._charge_(), iterates over every amino acid in self.protein to calculate the molecular weight of the protein.\n",
    "\n",
    "- How did you implement the charge method? how is pH given to charge by the pI method ?\n",
    "\n",
    "in self._charge_(), the first thing i do is initialize counters for the 2 sides of the net charge equation: the positive and negative charge sides. Then as indicated by the summation symbol, I iterate over the amino acids and check if they are a  charged amino acid, incrementing their respective charge type by the value of the summation equation (10^pKa / (10^pKa + 10^pH)). pH is stored as an iterator variable in the range [0, 1400] with the iterator being divided by 100 for the effective range to be [0.00, 14.00] (or [0.00, 14.01)). This value is then passed as an argument to self._charge_() within self.pI().\n",
    "\n",
    "- How did you iterate across the range of pH in order to get 2 decimal ponts of precision ( 7.16, for example )\n",
    "\n",
    "As described above, pH is iterated through in a linear fashion by using a range of [0, 1401), then dividing the iterator by 100 to set the range effectively to [0, 14.01).\n",
    "\n",
    "- How did you calculate massExtinction coefficient without having to redo your work from molarExtinction\n",
    "\n",
    "I divided the term self.molarExtinction() (which computes in runtime and that result is passed in the method's place) by the molecular weight if applicable, returning 0.0 otherwise. By calling self.molarExtinction() within self.massExtinction(), I save myself the effort of having to rewrite the method's code while still achieving the same effect.\n",
    "\n",
    "- How did you make use of the many dictionaries that are given in order to avoid having to build them from scratch ?\n",
    "\n",
    "I made use of both their keys and values, keys being used not only to access the values but to validate whether an amino acid is a valid amino acid (by checking if a given value is in the list of keys for self.aa2mw). The same type of technique was used for validating if an amino acid was charged or not in self._charge_()."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein Param"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Amino Acids: 14\n",
      "Molecular Weight: 1499.7\n",
      "molar Extinction coefficient: 5625.00\n",
      "mass Extinction coefficient: 3.75\n",
      "Theoretical pI: 14.00\n",
      "Amino acid composition:\n",
      "\tA = 21.43%\n",
      "\tC = 0.00%\n",
      "\tD = 7.14%\n",
      "\tE = 0.00%\n",
      "\tF = 0.00%\n",
      "\tG = 0.00%\n",
      "\tH = 0.00%\n",
      "\tI = 0.00%\n",
      "\tK = 14.29%\n",
      "\tL = 7.14%\n",
      "\tM = 0.00%\n",
      "\tN = 7.14%\n",
      "\tP = 7.14%\n",
      "\tQ = 0.00%\n",
      "\tR = 0.00%\n",
      "\tS = 7.14%\n",
      "\tT = 7.14%\n",
      "\tV = 14.29%\n",
      "\tW = 7.14%\n",
      "\tY = 0.00%\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "# Name: Akash Pandit\n",
    "# Group Members: Preeti Rajarathinam(prrajara), Delson Hays (dhhays), Natalie Cellucci(ncellucc), Riya Tiloda (rtiloda)\n",
    "\n",
    "from numpy import inf  # used in ProteinParam.pI()\n",
    "\n",
    "\"\"\"\n",
    "This program defines a class ProteinParam which is built around a string of amino acids represented as single characters.\n",
    "Many operations can be performed on these, as listed by the below methods\n",
    "\n",
    "self.aaCount() - find the number of amino acids in the input string.\n",
    "\n",
    "self.pI() - finds the isoelectric point, the pH value where the protein is most electrically neutral.\n",
    "\n",
    "self.aaComposition() - returns a dictionary mapping each unique amino acid in the protein to its count.\n",
    "\n",
    "self._charge_(pH) - an internal method that calculates the total electric charge of the protein at a specific pH.\n",
    "\n",
    "self.molarExtinction(oxidizing) - calculates the molar extinction coefficient for the protein. In an oxidizing environment, \n",
    "    cysteine is left out of the calculations as its present in cystine residues which do not contribute. In a reducing \n",
    "    environment, this does not occur and cysteine remains in the molar extinction coefficient calculations.\n",
    "\n",
    "self.massExtinction(oxidizing) - divides the molar extinction value from self.molarExtinction by the molar mass to get the\n",
    "    protein's mass extinction value. The oxidizing parameter value is passed to self.molarExtinction().\n",
    "\n",
    "self.molecularWeight() - calculates the molecular weight of the protein.\n",
    "\n",
    "SAMPLE INPUT:\n",
    "VLSPADKTNVKAAW\n",
    "\n",
    "SAMPLE OUTPUT:\n",
    "Number of Amino Acids: 14\n",
    "Molecular Weight: 1499.7\n",
    "molar Extinction coefficient: 5500.00\n",
    "mass Extinction coefficient: 3.67\n",
    "Theoretical pI: 9.88\n",
    "Amino acid composition:\n",
    "A = 21.43%\n",
    "C = 0.00%\n",
    "D = 7.14%\n",
    "E = 0.00%\n",
    "F = 0.00%\n",
    "G = 0.00%\n",
    "H = 0.00%\n",
    "I = 0.00%\n",
    "K = 14.29%\n",
    "L = 7.14%\n",
    "M = 0.00%\n",
    "N = 7.14%\n",
    "P = 7.14%\n",
    "Q = 0.00%\n",
    "R = 0.00%\n",
    "S = 7.14%\n",
    "T = 7.14%\n",
    "V = 14.29%\n",
    "W = 7.14%\n",
    "Y = 0.00%\n",
    "\n",
    "This program also includes a driver function main() to build a ProteinParam object and output\n",
    "the results of the object's methods.\n",
    "\"\"\"\n",
    "\n",
    "class ProteinParam:\n",
    "    \"\"\"class containing multiple data tables and methods for parsing strings of amino acids in their single character form\"\"\"\n",
    "# These tables are for calculating:\n",
    "#     molecular weight (aa2mw), along with the mol. weight of H2O (mwH2O)\n",
    "#     absorbance at 280 nm (aa2abs280)\n",
    "#     pKa of positively charged Amino Acids (aa2chargePos)\n",
    "#     pKa of negatively charged Amino acids (aa2chargeNeg)\n",
    "#     and the constants aaNterm and aaCterm for pKa of the respective termini\n",
    "#  Feel free to move these to appropriate methods as you like\n",
    "\n",
    "# As written, these are accessed as class attributes, for example:\n",
    "# ProteinParam.aa2mw['A'] or ProteinParam.mwH2O\n",
    "\n",
    "    aa2mw = {  # defines each amino acid by single character symbol with its molecular weight in daltons or grams per mole\n",
    "        'A': 89.093,  'G': 75.067,  'M': 149.211, 'S': 105.093, 'C': 121.158,\n",
    "        'H': 155.155, 'N': 132.118, 'T': 119.119, 'D': 133.103, 'I': 131.173,\n",
    "        'P': 115.131, 'V': 117.146, 'E': 147.129, 'K': 146.188, 'Q': 146.145,\n",
    "        'W': 204.225,  'F': 165.189, 'L': 131.173, 'R': 174.201, 'Y': 181.189\n",
    "        }\n",
    "\n",
    "    mwH2O = 18.015  # defines the molecular weight of water in daltons or grams per mole\n",
    "    aa2abs280= {'Y':1490, 'W': 5500, 'C': 125}  # defines absorption values for Y, W,and C\n",
    "\n",
    "    aa2chargePos = {'K': 10.5, 'R':12.4, 'H':6}  # defines the pKa of positively charged amino acids\n",
    "    aa2chargeNeg = {'D': 3.86, 'E': 4.25, 'C': 8.33, 'Y': 10}  # defines the pKa of negatively charged amino acids\n",
    "    aaNterm = 9.69  # pKa of N-terminus\n",
    "    aaCterm = 2.34  # pKa of C-terminus\n",
    "\n",
    "    def __init__ (self, protein):\n",
    "        \"\"\"constructor for ProteinParam, initializes the input protein and its composition in a dict\"\"\"\n",
    "        self.protein = ''.join([aa for aa in protein if aa in self.aa2mw.keys()]).upper()\n",
    "        # defines the protein as the input string with all erroneous characters removed\n",
    "        self.aaComposition_ = {aa : self.protein.count(aa) for aa in self.aa2mw}\n",
    "        # generates dict of key:value pairs in aa2mw if those keys are in proteins\n",
    "\n",
    "    def aaCount(self) -> int:\n",
    "        \"\"\"gets the length of the purified protein parameter\"\"\"\n",
    "        return len(self.protein)\n",
    "\n",
    "    def pI (self):\n",
    "        \"\"\"finds the isoelectric value for a given pH. \"\"\"\n",
    "        isoelectric = inf\n",
    "        for pH in range(0, 1401):\n",
    "            pH /= 100  # brings it down to the scale of 0.00 to 14.00\n",
    "            newCharge = abs(self._charge_(pH))  # we want closest to 0, so measure distance from 0 with abs()\n",
    "            if newCharge < isoelectric: \n",
    "                isoelectric = newCharge\n",
    "                return_pH = pH\n",
    "        if return_pH:\n",
    "            return return_pH\n",
    "        raise AttributeError(\"For some reason, return_pH was never executed. Investigate this.\")\n",
    "\n",
    "    def aaComposition(self) -> dict:\n",
    "        \"\"\"returns a dictionary of the amino acid composition of the protein with 1 letter shortened amino acids as keys\n",
    "        and their molecular weight as their value\"\"\"\n",
    "        return self.aaComposition_\n",
    "\n",
    "    def _charge_ (self, pH):\n",
    "        \"\"\"Implements the equation to calculate the net charge on an amino acid\"\"\"\n",
    "        posCharge, negCharge = 0, 0\n",
    "\n",
    "        for aa,pKa in self.aa2chargePos.items(): # Calculate positive charge\n",
    "            posCharge += self.aaComposition_[aa] * (10**pKa) / (10**pKa + 10**pH)\n",
    "            \n",
    "        for aa,pKa in self.aa2chargeNeg.items(): # Calculate negative charge\n",
    "            negCharge += self.aaComposition_[aa] * (10**pH) / (10**pKa + 10**pH)\n",
    "        \n",
    "        posCharge += 10**self.aaNterm / (10**self.aaNterm + 10**pH) # increments n terminus charge\n",
    "        negCharge += 10**pH / (10**self.aaCterm + 10**pH) # increments c terminus charge\n",
    "        \n",
    "        return posCharge - negCharge # returns net charge\n",
    "\n",
    "\n",
    "    def molarExtinction(self, oxidizing=True) -> float:\n",
    "        \"\"\"Implements the molar extinction equation, with a base True case to account for oxidative conditions and if false, reductive conditions\"\"\"\n",
    "        molarExt = 0\n",
    "        for aa in self.aa2abs280.keys():\n",
    "            if oxidizing and aa != 'C':\n",
    "                molarExt += self.aaComposition()[aa] * self.aa2abs280[aa]\n",
    "            elif not oxidizing:\n",
    "                molarExt += self.aaComposition()[aa] * self.aa2abs280[aa]\n",
    "\n",
    "        return molarExt\n",
    "\n",
    "\n",
    "    def massExtinction(self, oxidizing=True) -> float:\n",
    "        \"\"\"divides the molar extinction by molecular weight to get the mass extinction\"\"\"\n",
    "        myMW = self.molecularWeight()\n",
    "        return self.molarExtinction(oxidizing=oxidizing) / myMW if myMW else 0.0\n",
    "\n",
    "    def molecularWeight(self) -> float:\n",
    "        \"\"\"\n",
    "        Formula: MW(H2O) + Sum for all inputted amino acids(N(aa) * (MW(aa) - MW(H2O)))\n",
    "        \"\"\"\n",
    "        molWeight = self.mwH2O  # initialize mol weight w/ 1 water\n",
    "        for aa in self.protein: \n",
    "            molWeight += (self.aa2mw[aa] - self.mwH2O)  \n",
    "            # adds the difference of the amino acid mol. weight and the mol. weight of water to molWeight\n",
    "        return molWeight\n",
    "\n",
    "# Please do not modify any of the following.  This will produce a standard output that can be parsed\n",
    "    \n",
    "import sys\n",
    "def main():\n",
    "    inString = input('protein sequence?')\n",
    "    while inString :\n",
    "        myParamMaker = ProteinParam(inString)\n",
    "        myAAnumber = myParamMaker.aaCount()\n",
    "\n",
    "        print (\"Number of Amino Acids: {aaNum}\".format(aaNum = myAAnumber))\n",
    "        print (\"Molecular Weight: {:.1f}\".format(myParamMaker.molecularWeight()))\n",
    "        print (\"molar Extinction coefficient: {:.2f}\".format(myParamMaker.molarExtinction()))\n",
    "        print (\"mass Extinction coefficient: {:.2f}\".format(myParamMaker.massExtinction()))\n",
    "        print (\"Theoretical pI: {:.2f}\".format(myParamMaker.pI()))\n",
    "        print (\"Amino acid composition:\")\n",
    "        \n",
    "        if myAAnumber == 0 : myAAnumber = 1  # handles the case where no AA are present \n",
    "        \n",
    "        for aa,n in sorted(myParamMaker.aaComposition().items(), key= lambda item:item[0]):\n",
    "            print (\"\\t{} = {:.2%}\".format(aa, n/myAAnumber))\n",
    "    \n",
    "        inString = input('protein sequence?')\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "nteract": {
     "transient": {
      "deleting": false
     }
    }
   },
   "source": [
    "# Inspection results\n",
    "Who participated in your code inspection? What did they suggest? How was the inspection valuable to you or to your team ?\n",
    "\n",
    "Delson, Natalie, and Preeti participated in my code inspection, with their suggestions and my responses below. Personally, I find these inspections to be incredibly valuable, as seen by the many bugs my teammates caught below. Their keen eyes remind me to practice more diligence while I work."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Delson Hays (dhhays) - The code is very concise and you have lots of annotations, which is always good. However, I'm getting a few discrepancies between the template output and this one. It seems like for aaComposition, your dictionary only contains the first amino acid (inputting KCEHRYD, for instance, just returns K = 100.0%). Similarly, pI always returns 0.0, and based on some strange results im getting for molarExtinction and molecularWeight (Y and YYYYY have the same values), I think the underlying issue may be that your methods are only taking the first item in your amino acid sequence.\n",
    "\n",
    "RESPONSE: caught the bug, thanks :)\n",
    "\n",
    "Natalie Cellucci(ncellucc) - Everything is spaced out nicely and reads well. The outputs match the expected. All your comments make sense and are concise. My suggestion would be to change the pH range from (0, 1400) to (0, 1401) so it iterates over the full range of possible pH values. I also noticed that your pI function returns the lowest possible charge, instead of the pH for the lowest charge. The isoelectric point itself is a pH, but it looks like it was defined as a charge.\n",
    "\n",
    "RESPONSE: noted, charge updated with new range & proper output for pI!\n",
    "\n",
    "Preeti Rajarathinam(prrajara) - All your comments and variable names  are very thorough and everything runs. However, idk if its a me problem but whenever I run your code it gives me slightly different values than they should be for the coefficients. for reference this is what i get. In his the values are 5500 and 3.67\n",
    "\n",
    "RESPONSE: noted, looked into this for a while and realized when I implemented the extra credit, I included the cysteine by default in my calculations when it should have been excluded by default. That's what led to the slightly different values for the coefficients. Thank you!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
