{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find Unique\n",
    " - Deliverables:\n",
    "     - findUnique.py - 50 points ( We will test this code)\n",
    "     - Notebook with Inspection materials and Results ( 5 pts )\n",
    "     - required input from STDIN\n",
    "     - required output to STDOUT\n",
    "     - output must be sorted by tRNA header\n",
    "\n",
    "## Sets and tRNA\n",
    "\n",
    "Mitochondrial tRNA are encoded in the mitochondrial genome and account for ~10% of the coding space, yet over 50% of mitochondrial genomic disease have their origin in this molecule. These molecules are transcribed, processed, modified, and ultimately folded to produce these mature adapters between the mitochondrial transcription and translation processes. Defects in mt.tRNA maturation reflect additional disease states .. if only we could identify those mutations. We are working on such a device, though we first need to use it to identify abundance of these molecules in a mixed population. Are there any unique subsequences among the 22 mt.tRNA that can be used as \"tags\"? If so, we can count those tags to assess abundance.\n",
    "\n",
    "## Assignment\n",
    "\n",
    "Write a python command line program that reads a file of fasta sequences from STDIN, finds the unique subsequences that occur in each single tRNA such that no members of this set occur among any of the other tRNA sets. Each of those 22 sets should be minimized, such that no member of that unique subsequence set is a substring of any other member of this set. [ Unique and Essential]\n",
    "\n",
    "As an example, let's say that both ACG and AAACGA are in a unique set. Since ACG is a substring of AAACGA we would remove AAACGA. [ ACG is Essential ]\n",
    "\n",
    "Use Python sets for this assignment[__required__]. Not only will your code be smaller, but it will be more likely to work. The union, intersection and difference operators will be quite useful.\n",
    "\n",
    "## Rough design plan...\n",
    "0) Consider a tRNA as an object. In thi scase we would then have 22 objects that would need to be able to communicate with each other.\n",
    "\n",
    "1) compute the set of all substrings from each tRNA sequence. [powerset] I will refer to a set of substrings as a tRNAset.\n",
    "\n",
    "2) for each tRNAset, compute the union of all other tRNA sets, and then remove that union from the current tRNAset. Notice that this union operation finds all of the substrings from all other tRNA. If any of those are present in your current tRNA, then they are not unique ! [ Unique]\n",
    "\n",
    "3) for each unique tRNAset, it now contains the truly unique ones, along with any extensions of that subsequence. If, for example, it was found that G only occurred in a single tRNA, then adding an A onto that G must also be unique because it has a G in it. We only want the minimal form.. G. [ Essential]\n",
    "\n",
    "4) Remove spaces from the header line before sorting and printing. This will make your output a little prettier.\n",
    "\n",
    "5) make sure to remove any alignment characters from your initial sequences. These characters are periods(.), underscores(\\_), or dashes(-) . You will find many new characters in the sequence other than {ACGU} - leave these in place, they are modified bases.\n",
    "\n",
    "## Report\n",
    "Print a report that contains items as follows. \n",
    "\n",
    " - Line 1: the tRNA name\n",
    " - Line 2: the tRNA sequence ( with the alignment characters removed)\n",
    " - lines 3-80 or so, each unique element.\n",
    "\n",
    "These unique elements need to be ordered by their starting position in the tRNA sequence, and should be spaced over so they are directly under that subsequence in the original tRNA. This looks like an alignment, but you can find where it belongs by using the string.find() method. Include dots in all positions to the left, serving as alignment guides for your reader. [ see sample output below ]\n",
    "\n",
    "Do this for all of the 22 tRNA sequences.\n",
    "\n",
    "Print the tRNA out as above, sorted by the header line.  \n",
    "\n",
    "## Hints:\n",
    "\n",
    "__use sets and the set operators - this is required !__\n",
    "\n",
    "your final code will be under 100 lines.\n",
    "\n",
    "Do most of your coding using methods defined in a class and then instantiated as objects. \n",
    "\n",
    "The sequences include characters that are just alignment characters. They are not part of the sequence and must be removed. [-\\_\\.] are alignment characters. \n",
    "\n",
    "When removing items from a tRNAset, don't do this while iterating through that set. Also, when building a unique set you will need the original contents of all other tRNAsets. So.. build a new set to keep the unique contents, or keep track of the elements you intend to delete later. Notice that you build the union of all other tRNA, and this happens 22 times - these unions are all distinct from each other. Example, consider 4 sets, A,B,C,D.  we would compute the set B, C, D to use against set A, and we would compute the set A, C, D to use against B. A and B need to not change while we are doing this computation.\n",
    "\n",
    "There are cases where a unique element is present multiple times in a single tRNA. This sequence is unique to this tRNA and should be reported in its multiple locations ( same sequence on multiple lines of the output).\n",
    "\n",
    "This is a command line program, though it does not have any optional parameters. You really don't need the commandline class. Your input comes from STDIN and output goes to STDOUT. Use the FastaReader for input and print statement for output. If you are going to use jupyter to develop your code, you can add a filename to fastareader for your testing.\n",
    "\n",
    "## Submission\n",
    "\n",
    "Submit code using Canvas. As always, work with your group and write your own code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "# Name: Akash Pandit (aspandit)\n",
    "# Group: None\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "    Input for STDIN (in order to compile):\n",
    "    python3 findUnique.py <bos-tRNA-7.fa >bos-tRNA-7-output.txt\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "from sequenceAnalysis import FastAreader\n",
    "\n",
    "class findUnique:\n",
    "    def __init__(self):\n",
    "        \"\"\"Reads tRNA from the inputted Fasta file. Adds the power set of each sequence to a list\"\"\"\n",
    "        self.pSets, self.uniques, self.headerDict = [], [], {}\n",
    "\n",
    "    def genPSets(self, sequence: str) -> None:\n",
    "        \"\"\"Appends all unique tRNA substrings (a powerset)\"\"\"\n",
    "        powerSets = set()\n",
    "        for index in range(len(sequence)):\n",
    "            size = len(sequence)\n",
    "            while size > index:\n",
    "                powerSets.add(sequence[index:size])\n",
    "                size -= 1\n",
    "\n",
    "        self.pSets.append(powerSets)\n",
    "\n",
    "    def findUniques(self):\n",
    "        \"\"\"\n",
    "        Looks for duplicate tRNA sequences and removes from all but the appriopiate set to\n",
    "        obtain a unique set of tRNA subsequences.\n",
    "        \"\"\"\n",
    "        for pSet in self.pSets:\n",
    "            union = set()\n",
    "            copyList = self.pSets.copy()\n",
    "            copySet = pSet.copy()  # Make duplicate of the possible power set\n",
    "            copyList.remove(copySet)  # Remove power set from list.\n",
    "\n",
    "            for powerSet in copyList: # removes the duplicate elements between union and copySet\n",
    "                union = union.union(powerSet)  # The union of all the other tRNA sets.\n",
    "\n",
    "            copySet.difference_update(union)  # Update the copySet, removing elements found in union.\n",
    "\n",
    "            newSet = copySet.copy()\n",
    "            for s1 in copySet:  # checks prior strings to identify more possible substrings\n",
    "                uniqueSet = copySet.copy()\n",
    "                uniqueSet.remove(s1)\n",
    "                for s2 in uniqueSet:\n",
    "                    if s1 in s2:\n",
    "                        if len(s1) < len(s2):\n",
    "                            newSet.discard(s2)  # Gets rid of the larger substrings.\n",
    "\n",
    "            self.uniques.append(newSet)\n",
    "\n",
    "    def print(self) -> None:\n",
    "        \"\"\"Prints to STDOUT (CL in this case)\"\"\"\n",
    "        for index in range(0,len(self.headerDict)):\n",
    "            header, sequence = self.headerDict[index]\n",
    "            print(header)\n",
    "            print(sequence)\n",
    "            for i in range(len(sequence)): \n",
    "                for substr in self.uniques[index]:\n",
    "                    if substr == sequence[len(substr)]:\n",
    "                        print('.'*i + substr)\n",
    "\n",
    "def main():\n",
    "    \"\"\"Runs above code to find and print header, sequence and substrings from the fasta file\"\"\"\n",
    "\n",
    "    # init objects & counter var\n",
    "    tRNA = findUnique()\n",
    "    fastaFile = FastAreader()\n",
    "    count = 0\n",
    "\n",
    "    # generate psets & increment counter\n",
    "    for header, sequence in fastaFile.readFasta():\n",
    "        filteredSequence = sequence.replace('-', '').replace('_', '')  # makes sure only valid characters are being used\n",
    "        tRNA.headerDict[count] = [header, filteredSequence]\n",
    "        tRNA.genPSets(filteredSequence)  # generate & append pset to self.psets\n",
    "        count += 1\n",
    "\n",
    "    # find & print unique seqs\n",
    "    tRNA.findUniques()\n",
    "    tRNA.print()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tRNA|Lys|∃UU|Bostaurus|mitochondrial\n",
    "CACUAAGA\"LCUAUAUAGCACPAACCU∃UU6AGUUAGAGAUUGAGAGCCAU\"UACUCUCCUUGGUGACCA\n",
    "CACU\n",
    ".ACUA\n",
    "...UAA\n",
    "....AAG\n",
    ".......A\"\n",
    ".........L\n",
    "..........CUAU\n",
    "............AUAU\n",
    "..............AUAG\n",
    "...............UAGC\n",
    ".................GCA\n",
    ".....................P\n",
    "......................AAC\n",
    ".......................ACCU\n",
    "...........................∃\n",
    "..............................6\n",
    "...............................AGU\n",
    "................................GUU\n",
    ".................................UUA\n",
    "..................................UAGA\n",
    "...................................AGAGA\n",
    "......................................GAU\n",
    ".......................................AUU\n",
    "........................................UUGA\n",
    ".........................................UGAG\n",
    "..........................................GAGAG\n",
    "............................................GAGC\n",
    "..............................................GCC\n",
    "................................................CAU\n",
    "..................................................U\"\n",
    "...................................................\"U\n",
    "....................................................UAC\n",
    ".....................................................ACUC\n",
    ".......................................................UCU\n",
    ".........................................................UCC\n",
    "...........................................................CUU\n",
    "..............................................................GG\n",
    "...............................................................GUG\n",
    ".................................................................GAC\n",
    "..................................................................ACCA"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
